Show IPC profile markers

jimporter opened the pull request:
This resolves #2164.

julienw commented:
hey @jimporter, do you need some feedback on this already, or do you plan some more work before this?
Thanks!

jimporter commented:
@julienw I do plan to keep working on this, but some feedback on the profile processing bits would be much appreciated! I'll be out Wed-Fri, but I'll get back to working on this in earnest on Monday.

gregtatum commented on src/profile-logic/convert-markers.js:
```diff
@@ -127,3 +130,49 @@ export function upgradeGCMajorMarker_Processed8to9(
   }
 }
+
+export function processIPCMarkers(threads: Array<Thread>) {
```
I'm wondering if this work could happen in a selector instead of during profile processing. Profile processing is "forever", and if there is a bug, then it is quite hard to go back and fix. Selectors run every time you load the profile, and compute derived data.

Here is where we derive the markers:

https://github.com/firefox-devtools/profiler/blob/f9bf24485b60c9a9ccd4cfcb90f22ea129097a6d/src/selectors/per-thread/markers.js#L82-L89

It could be possible to wire in the information to this selector, or create a new one outside of the per-thread infrastructure. Happy to chat real-time about this if you want.

jimporter commented:
@gregtatum Ok, I updated this patch to do the IPC marker correlation via a selector. It's a big mess right now, but I wanted to make sure I was headed in the right direction with this before I spent too long cleaning things up.

Note: currently, the change to using a selector is just a commit on top of the previous one; that's just so I don't lose the prior commit while I'm figuring out how to do this the right way. If the current code looks basically good, I'll squash the commits and get rid of the bad old version.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -459,6 +461,30 @@ export function deriveMarkersFromRawMarkerTable(
       }
 
+      case 'IPC': {
+        const times = ipcCorrelations[[threadId, i]];
```
I'm a bit confused by the `ipcCorrelations` structure, and how it's working with taking a tuple like `[threadId, i]`. Also, it's best to avoid small allocations if possible in hot code, which all this marker iteration is going to be hot.

gregtatum replied:
Oh wow, I guess this runs `toString()` on the array, and uses that as a key. I'd prefer a more straightforward approach with string concatenation. This code will be hot enough to justify writing C-like code as much as possible.

jimporter replied:
Yeah, this was originally going to be a `Map` with a pair as the key, but I forgot that comparison of two JS arrays compares their memory locations, not their values. So for the time being, I figured the easiest change was to use an `Object` instead and worry about doing it right later. :)

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -1056,6 +1082,13 @@ export function getMarkerFullDescription(marker: Marker) {
         description += ` — ${data.name}`;
         break;
+      case 'IPC':
+        description = `[${data.thispid}${data.direction ? '<-' : '->'}${
```
Some comments showing examples of what this spits out would be helpful, the line right now is hard to visually parse.

jimporter replied:
I'm planning on reworking this anyway, since the majority of the information here is debug info I was using to check that I'm correlating IPC markers correctly. If it's still complicated, I'll definitely add a comment though!

gregtatum commented on src/selectors/per-thread/markers.js:
```diff
@@ -82,9 +84,11 @@ export function getMarkerSelectorsPerThread(threadSelectors: *) {
     _getFirstSampleTime,
     _getLastSampleTime,
     ProfileSelectors.getProfileInterval,
+    ProfileSelectors.getIPCCorrelations,
```
👍

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
   return selectors;
 };
 
+const getThreadsForPIDs = createSelector(
```
These will need type information, and code comments.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+const getThreadsForPIDs = createSelector(
+  getThreads,
+  (threads) => {
+    let pid2threads = {};
```
This name is confusing to me, is it supposed to be `pidToThreads`?  It's also more common to the codebase to use the `Map` data structure, although this is functionally the same.

jimporter replied:
Yeah, it's a mapping from PIDs to thread objects. Using "2" for "to" in symbol names is just a (possibly bad) habit I picked up from somewhere...

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+              dstData.messageseqno === srcData.messageseqno &&
+              dstData.messageid === srcData.messageid &&
+              dstData.direction === 1) {
+            return [thread, i, dstData];
```
A named object here would be more maintainable for future readers, e.g. `{ thread, i, dstData }`

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+        throw new Error('recipient not found');
+      }
+      for (let thread of pid2threads[srcData.otherpid]) {
+        for (let i = 0; i < thread.markers.length; i++) {
```
We deal with lots of indexes in the profiler, so I usually find it's helpful to name the indexes as I use them, especially as this one is getting passed into another data structure. I would rename `i` to `markerIndex`.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+          }
+        }
+      }
+      throw new Error('recipient not found');
```
This could use a more helpful error message, with a full sentence.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+      throw new Error('recipient not found');
+    }
+
+    let result = {};
```
Could this be named `IPCCorrelations`? It's also nice to do an inline type definition here, as Flow will surface errors better if the type is defined at its creation, along with on the function signature.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -139,6 +139,64 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+          const [dstThread, dstIndex, dstData] = findRecipientMarker(srcData);
+          const pairData = {startTime: srcData.startTime,
+                            endTime: dstData.startTime};
+          result[[thread.tid, i]] = pairData;
```
I don't think this will work by having a tuple as a key to an object?

jimporter commented:
@gregtatum I reworked the selector to build a lookup table for the recipient IPC markers instead of the PID->thread mapping. This takes more memory, but it should change the selector from (roughly) *O(n^2)* to *O(n)*. There's still a bunch more to do on improving the selector (e.g. making the `Map` key generation more-performant), but I'm going to do that after some fixes to the Gecko patch.

jimporter commented:
Here's a profile I generated that you can use to try this PR out: https://perfht.ml/2NmOqzf

(Just download it and open it up in a local instance of the profiler frontend. I'm not sure if there's an easier way to let people try this...)

jimporter commented:
@gregtatum, I think all that's left here is to add some tests, and possibly tweak the UI based on feedback from people trying this out. Aside from that, does this look about right?

jimporter commented on src/selectors/profile.js:
```diff
@@ -139,6 +140,103 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+  }
+
+  _makeKey(tid: number | void, index: number) {
+    return [tid, index].toString();
```
That makes sense and should hopefully be a little nicer performance-wise.

jimporter commented on src/selectors/profile.js:
```diff
@@ -139,6 +140,103 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
+        if (dstMarkers.has(key)) {
+          // FIXME: When we have a more-reliable way of identifying correlated
+          // IPC markers, we won't have to worry about this.
+          console.warn('[squib] duplicate message found', key);
```
Oh, right. I forgot about this bit. I'm going to try and remove the need for this entirely if I can get the backend to behave a bit better...

jimporter replied:
I've pretty much convinced myself that this can't ever happen now, unless the profile is corrupted or something changes substantially in Gecko. I think it's worth leaving a warning here just in case, or maybe even upgrading it to an exception...

jimporter commented on src/profile-logic/marker-data.js:
```diff
@@ -1056,6 +1091,11 @@ export function getMarkerFullDescription(marker: Marker) {
         break;
+      case 'IPC':
+        description = `${data.messagetype} (${
+          data.direction === 'sending' ? '🡆' : '🡄'
```
Those are supposed to be right/left arrows, respectively. I'm not sure that's the best way to show this information though...

jimporter commented on src/profile-logic/marker-styles.js:
```diff
@@ -152,6 +152,17 @@ export const markerStyles: MarkerStyles = {
     background: colors.BLUE_50,
   },
+
+  IPCOut: {
```
Profiling IPC messages is off by default (for now). I'm open to better ways to expose this on the timeline though. I'm not sure what people would like the most... I'm open to ideas!

jimporter commented:
@gregtatum Ok, I think I've resolved everything except the visual issues and the tests. Thanks for taking a look at this!

If you have any thoughts about how I can improve the UI further (e.g. putting these markers somewhere special instead of in the generic marker timeline), just let me know. I'll keep experimenting with some alternate UIs on my end as well.

jimporter commented:
@gregtatum I think I've addressed all the issues here. I do plan to work on a better UI for the timeline later, though I didn't want this to get too bogged down in UI tinkering before we have anyone else trying to use this. Adding a special timeline track for IPC messages would probably be good, though I think the specifics of what it would look like depend on people's needs...

jimporter commented:
Here's a new deploy preview with an updated profile (I changed the format of the IPC markers): https://deploy-preview-2172--perf-html.netlify.com/public/3e3be56561ef3a3ebc089fdf3a77006dac578c64

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -243,10 +255,108 @@ export function extractMarkerDataFromName(
+ * that we can share data between the two during profile processing.
+ */
+export function correlateIPCMarkers(threads: Thread[]): IPCMarkerCorrelations {
+  // Create a unique ID constructed from the source PID, destination PID,
```
I would check `stringTable.hasString('IPC')` here, so we can bail out early if there are no IPC markers. It could be wrong, but I tested a few place and it was a pretty good "cheap" check.

gregtatum replied:
You handled this while i was reviewing :)

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -243,10 +255,114 @@ export function extractMarkerDataFromName(
+  // destination marker (if any).
+  const correlations = new IPCMarkerCorrelations();
+  for (const thread of threads) {
+    if (!thread.stringTable.hasString('IPC')) {
```
Maybe provide a quick comment here to add context. Something like:

> Only check for IPC markers if the string table has the string "IPC". This is an easy optimization to avoid looping through all of the markers if we don't have to.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -598,22 +605,30 @@ export function deriveMarkersFromRawMarkerTable(
+        const pairData = ipcCorrelations.get(threadId, i);
+        let otherThreadName;
+        if (pairData) {
+          otherThreadName = `${friendlyThreadNames.get(
```
I think this could be computed in the Tooltip view layer instead of here. This way we don't have to create a bunch of strings that may never be viewed.

https://github.com/firefox-devtools/profiler/blob/master/src/components/tooltip/Marker.js#L232

I think this component could also just compute the friendly thread name directly using the threads and current thread, rather than using the selector `getFriendlyThreadNames`.

jimporter replied:
This is a bit more complex since I wouldn't be passing the "threads and current thread", but rather the threads and *target* thread. That's doable (probably), but it means looking up the target thread by its TID and then calling `ProfileData.getFriendlyThreadName`, once per IPC marker.

I'm concerned that this would end up being considerably more work than just making a "TID -> friendly name" mapping one time and caching it as a selector. I also wasn't sure how to pipe the extra data I needed to generate a friendly name all the way to the view functions, especially for `getMarkerFullDescription`, which only takes a marker as input.

Maybe it would make sense to take the existing `getFriendlyThreadName` selector and have it call my `getFriendlyThreadNames` selector though? I'm not 100% sure on what the perf difference would be there...

gregtatum replied:
> I'm concerned that this would end up being considerably more work than just making a "TID -> friendly name" mapping one time and caching it as a selector

Ah, I didn't realize the complexity here! This is fine then. Could you add a comment to the selector explaining this?

If you feel that this is the simpler implementation, then I'm fine moving forward. I just assumed the tooltip component would have access to all of the information it needed.

jimporter commented:
@gregtatum I came up with a simpler way to fetch the friendly thread name. Does this look ok?

jimporter commented:
@julienw Do you think you could take a quick look at the second commit in this PR while @gregtatum is out? I think it's pretty close to good, though I'm not happy with my change to `_getDefaultLocalTrackOrder`.

Maybe there's a better way to do what I want there. I'm really just trying to make sure that the IPC track goes below the corresponding thread in the UI so that it's easy to tell which IPC track belongs to which thread. I'm sure this will need tests and such too, but I'll do those once we're basically happy with the code.

We could also pull the second commit out into another PR so that this one can get landed sooner. It'd be nice to get this published soon so people can profile our IPC more easily!

gregtatum commented on src/components/timeline/Markers.js:
```diff
@@ -475,3 +475,21 @@ export const TimelineMarkersMemory = explicitConnect<
   component: TimelineMarkers,
 });
+
+export const TimelineMarkersIPC = explicitConnect<OwnProps, StateProps, {||}>({
```
Generally we try to add a bit of documentation for each component to provide a bit of context.

gregtatum commented on src/components/timeline/TrackIPC.css:
```diff
@@ -0,0 +1,8 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+.timelineMarkersIPC {
+  height: var(--markers-height, 25px);
```
Seems like we don't need to repeat the 25px here, as it'll be easy to get out of date.

gregtatum commented on src/components/timeline/TrackIPC.js:
```diff
@@ -0,0 +1,82 @@
+
+type State = {||};
+
+export class TrackIPCImpl extends React.PureComponent<Props, State> {
```
And also a short bit of documentation here would be good.

gregtatum commented on src/components/tooltip/Marker.js:
```diff
@@ -620,6 +620,15 @@ function getMarkerDetails(
+        tooltipDetails = (
+          <TooltipDetails>
+            <TooltipDetail label="Type">{data.messageType}</TooltipDetail>
+            <TooltipDetail label="Sync">{data.sync.toString()}</TooltipDetail>
```
I believe this is crashing in the current deploy previews. Is there an updated one?

jimporter replied:
Fixed! This was because I forgot to update the profile being used by the preview. Here's a new one: https://deploy-preview-2172--perf-html.netlify.com/public/dab7795ca9494375f62e4adfe7826e94633825aa

gregtatum commented on src/profile-logic/tracks.js:
```diff
@@ -56,11 +58,20 @@ const GLOBAL_TRACK_DISPLAY_ORDER = {
+    // that IPC tracks are grouped below the thread they correspond to.
+    const threadA = 'threadIndex' in tracks[a] ? tracks[a].threadIndex : -1;
+    const threadB = 'threadIndex' in tracks[b] ? tracks[b].threadIndex : -1;
+    if (threadA !== threadB) {
```
I'm a bit torn here on the implementation. The current metaphor for tracks is that each track is either a global track, or a local track. A global track is an individual process, or something like a screenshots. The local tracks are something that is going on in that process. Having the IPC tracks as a local track is somewhat confusing this breakdown.

Would it be possible to create 1 local track for the entire process's IPC? Then inside that track, each thread would have a separate breakdown.

jimporter commented:
Here's a new profile to test the UI with (this adds the `sync` attribute to IPC markers): https://deploy-preview-2172--perf-html.netlify.com/public/dab7795ca9494375f62e4adfe7826e94633825aa

jimporter commented:
@gregtatum Ok, I think this should be good now, at least as an initial version.
